guide/language-concepts/normal-ordering.markdown

```diff
@@ -19,22 +19,69 @@ You can override normal ordering in exceptional circumstances by making a
 promise in a class context and defining that class based on the outcome of
 another promise, or using the `depends_on` promise attribute.
 
-### Agent normal ordering
+## Agent normal ordering
 
 CFEngine tries to keep variable and class promises before starting to consider
-any other kind of promise. In this way, global variable and classes can be set.
+any other kind of promise. In this way, global variables and classes can be set.
 
 If you set variables based on classes that are determined by other variables,
 then you introduce an order dependence to the resolution that might be
 non-unique. **Since CFEngine starts trying to converge values as soon as
 possible, it is best to define variables in bundles before using them**, i.e. as
-early as possible in your configuration. In general it is wise to avoid
-class-variable dependency as much as possible.
+early as possible in your configuration. In order to make sure all global
+variables and classes are available early enough policy pre-evaluation step was
+introduced.
 
-CFEngine executes agent promise bundles in the strict order defined by the
+### Policy evaluation details
+
+Before exact evaluation of promises takes place first command line
+parameters are read and all classes defined using -D parameter are set. Next,
+environment detection takes place and hard classes are discovered.
+When environment detection is complete all the persistent classes are 
+loaded and a policy sanity check is performed using cf-promises.
+
+#### cf-promises policy validation step
+
+In this step policy is validated and classes and variables promises are evaluated. 
+Note that cached functions are executed here, and then again during the normal agent
+execution. Also policy validation is done using separate binary and any classes and
+variables are not preserved during validation step and normal agent run.
+
+#### Agent pre-evaluation step
+
+In order to support dynamic inputs and make sure all needed classes and
```

Comment:
I'm not sure if people know what "dynamic inputs" is. Perhaps say (i.e. expansion of variables in body common control inputs). In fact I think we should not say "dynamic inputs" at all - it's yet another concept that needs to be understood..